source/faucet/main.d

```diff
@@ -382,6 +382,15 @@ public class Faucet : FaucetAPI
         auto owned_utxo_rng = this.state.owned_utxos.byKeyValue()
             // do not pick already used UTXOs
             .filter!(pair => pair.key !in this.used_utxos);
+
+        auto owned_utxo_len = owned_utxo_rng.walkLength;
+        assert(owned_utxo_len >= 0);
+        if (owned_utxo_len <= 1)
+        {
+            logError("Insufficient UTXOs in storage. # of UTXOs: %s", owned_utxo_len);
+            throw new Exception(format("Insufficient UTXOs in storage. # of UTXOs: %s", owned_utxo_len));
+        }
+
         auto first_utxo = owned_utxo_rng.front;
```

Comment:
```suggestion
        auto first_utxo = owned_utxo_rng.take(1).array;
        if (first_utxo.length == 0)
        {
            logError("Insufficient UTXOs in storage. # of UTXOs: %s", owned_utxo_len);
            throw new Exception(format("Insufficient UTXOs in storage. # of UTXOs: %s", owned_utxo_len));
        } else {
           ....
        }
```

Reply:
Something like that ^

Reply:
Oh, I didn't think of the `array` call..?

Reply:
I was trying to avoid the `array` though.

Reply:
Why? It is only at most 1 in length.

Reply:
Yes, or 2 because `length == 1` should also throw.

Reply:
ok so you create an array with at most two elements.

Reply:
Also I thought if you use the last UTXO it will have change which will result in another UTXO. So you can use the last one.

Reply:
:thinking:

Reply:
we are apparently `Attempting to fetch the front of an empty filter`.

Reply:
Push your branch and I'll take a look.

Reply:
You probably didn't use array `opindex` (i.e. first_utxo[0] ) ?